rclpy/src/rclpy/subscription.cpp

```diff
@@ -0,0 +1,162 @@
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini subscription: %s",
+      rcl_get_error_string().str);
+  }
+  PyMem_Free(sub);
+}
+
+py::capsule
+subscription_create(
+  py::capsule pynode, py::object pymsg_type, std::string topic,
+  py::capsule pyqos_profile)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  auto msg_type = static_cast<rosidl_message_type_support_t *>(
+    rclpy_common_get_type_support(pymsg_type.ptr()));
+  if (!msg_type) {
+    throw py::error_already_set();
+  }
+
+  rcl_subscription_options_t subscription_ops = rcl_subscription_get_default_options();
+
+  if (!pyqos_profile.is_none()) {
+    if (0 != strcmp("rmw_qos_profile_t", pyqos_profile.name())) {
+      throw py::value_error("capsule is not an rmw_qos_profile_t");
+    }
+    auto qos_profile = static_cast<rmw_qos_profile_t *>(pyqos_profile);
+    subscription_ops.qos = *qos_profile;
+  }
+
+  // Use smart pointer to make sure memory is free'd on error
+  auto deleter = [](rclpy_subscription_t * ptr) {_rclpy_destroy_subscription(ptr);};
+  auto sub = std::unique_ptr<rclpy_subscription_t, decltype(deleter)>(
+    static_cast<rclpy_subscription_t *>(PyMem_Malloc(sizeof(rclpy_subscription_t))),
+    deleter);
+  if (!sub) {
+    throw std::bad_alloc();
+  }
+  sub->subscription = rcl_get_zero_initialized_subscription();
+  sub->node = node;
+
+  rcl_ret_t ret = rcl_subscription_init(
+    &(sub->subscription), node, msg_type,
+    topic.c_str(), &subscription_ops);
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_TOPIC_NAME_INVALID) {
+      std::string error_text{"Failed to create subscription due to invalid topic name '"};
+      error_text += topic;
+      error_text += "'";
+      throw py::value_error(error_text);
+    }
+    throw RCLError("Failed to create subscription");
+  }
+
+  PyObject * pysub_c =
+    rclpy_create_handle_capsule(sub.get(), "rclpy_subscription_t", _rclpy_destroy_subscription);
+  if (!pysub_c) {
+    throw py::error_already_set();
+  }
+  auto pysub = py::reinterpret_steal<py::capsule>(pysub_c);
+  // pysub now owns the rclpy_subscription_t
+  sub.release();
+
+  auto sub_handle = static_cast<rclpy_handle_t *>(pysub);
+  auto node_handle = static_cast<rclpy_handle_t *>(pynode);
+  _rclpy_handle_add_dependency(sub_handle, node_handle);
+  if (PyErr_Occurred()) {
+    _rclpy_handle_dec_ref(sub_handle);
+    throw py::error_already_set();
+  }
+
+  return pysub;
+}
+
+py::object
+subscription_get_logger_name(py::capsule pysubscription)
+{
+  auto sub = static_cast<rclpy_subscription_t *>(
+    rclpy_handle_get_pointer_from_capsule(pysubscription.ptr(), "rclpy_subscription_t"));
+  if (!sub) {
+    throw py::error_already_set();
+  }
+
+  const char * node_logger_name = rcl_node_get_logger_name(sub->node);
```

Comment:
This function appears to be a duplicate. I think `rclpy_subscription_get_logger_name` and `rclpy_publisher_get_logger_name` could just be deleted, and [`rclpy_get_node_logger_name` used instead](https://github.com/ros2/rclpy/blob/415717c870a21875e7552bf1bece8d1bef8f8544/rclpy/src/rclpy/_rclpy.c#L831).

Reply:
Dropped in d838d24.

Reply:
I dropped the publisher function in the publisher PR: c37603a

Reply:
It looks like CI is failing because Python code still calls it. It seems like the python publisher/subscription code doesn't have access to the node, so someone did it in C to get the node off the `rclpy_publisher/subscription_t` type.

It makes me wonder why the publisher or subscription need a method for loggers at all? I guess because this code doesn't have access to a node? https://github.com/ros2/rclpy/blob/415717c870a21875e7552bf1bece8d1bef8f8544/rclpy/rclpy/qos_event.py#L231

I'm fine with leaving the duplicate methods in, and adding an issue to fix it later.

Reply:
Restored in e00c660